src/native/eventpipe/ds-ipc-pal-socket.c

```diff
@@ -0,0 +1,1501 @@
+
+	ds_ipc_socket_t new_socket = DS_IPC_INVALID_SOCKET;
+	DS_ENTER_BLOCKING_PAL_SECTION;
+	new_socket = socket (ipc->server_address_family, SOCK_STREAM, 0);
+	DS_EXIT_BLOCKING_PAL_SECTION;
+	return new_socket;
+#endif
+	return DS_IPC_INVALID_SOCKET;
+}
+
+static
+inline
+ds_ipc_socket_t
+ipc_socket_create_tcp (DiagnosticsIpc *ipc)
+{
+#if defined(DS_IPC_PAL_AF_INET) || defined(DS_IPC_PAL_AF_INET6)
+#if defined(DS_IPC_PAL_AF_INET) && defined(DS_IPC_PAL_AF_INET6)
+	EP_ASSERT (ipc->server_address_family == AF_INET || ipc->server_address_family == AF_INET6);
+#else
+	EP_ASSERT (ipc->server_address_family == AF_INET);
+#endif
+
+	ds_ipc_socket_t new_socket = DS_IPC_INVALID_SOCKET;
+	DS_ENTER_BLOCKING_PAL_SECTION;
+	new_socket = socket (ipc->server_address_family, SOCK_STREAM, IPPROTO_TCP);
+	if (new_socket != DS_IPC_INVALID_SOCKET && ipc->mode == DS_IPC_CONNECTION_MODE_LISTEN) {
+		int option_value = 1;
+		setsockopt (new_socket, SOL_SOCKET, SO_REUSEADDR, (const char*)&option_value, sizeof (option_value));
+#ifdef DS_IPC_PAL_AF_INET6
+		if (ipc->is_dual_mode) {
+			option_value = 0;
+			setsockopt (new_socket, IPPROTO_IPV6, IPV6_V6ONLY, (const char*)&option_value, sizeof (option_value));
+		}
+#endif
+	}
+	DS_EXIT_BLOCKING_PAL_SECTION;
+	return new_socket;
+#endif
+	return DS_IPC_INVALID_SOCKET;
+}
+
+static
+inline
+ds_ipc_socket_t
+ipc_socket_create (DiagnosticsIpc *ipc)
+{
+#if defined(DS_IPC_PAL_AF_UNIX)
+	return ipc_socket_create_uds (ipc);
+#elif defined(DS_IPC_PAL_AF_INET) || defined(DS_IPC_PAL_AF_INET6)
+	return ipc_socket_create_tcp (ipc);
+#else
+	#error "Unknown address family"
+#endif
+}
+
+static
+inline
+int
+ipc_socket_close (ds_ipc_socket_t s)
+{
+	int result_close;
+	DS_ENTER_BLOCKING_PAL_SECTION;
+#ifdef HOST_WIN32
+	result_close = closesocket (s);
+#else
+	result_close = close (s);
+#endif
+	DS_EXIT_BLOCKING_PAL_SECTION;
+	return result_close;
+}
+
+static
+inline
+int
+ipc_socket_set_permission (ds_ipc_socket_t s)
+{
+#if defined(DS_IPC_PAL_AF_UNIX) && !defined(__APPLE__)
+	int result_fchmod;
+	DS_ENTER_BLOCKING_PAL_SECTION;
+	result_fchmod = fchmod (s, S_IRUSR | S_IWUSR);
+	DS_EXIT_BLOCKING_PAL_SECTION;
+	return result_fchmod;
+#else
+	return 0;
+#endif
+}
+
+static
+inline
+int
+ipc_socket_set_blocking (
+	ds_ipc_socket_t s,
+	bool blocking)
+{
+	int result = DS_IPC_SOCKET_ERROR;
+	DS_ENTER_BLOCKING_PAL_SECTION;
+#ifdef HOST_WIN32
+	u_long blocking_mode = blocking ? 0 : 1;
+	result = ioctlsocket (s, FIONBIO, &blocking_mode);
+#else
+	result = fcntl (s, F_GETFL, 0);
+	if (result != -1)
+		result = fcntl (s, F_SETFL, blocking ? (result & (~O_NONBLOCK)) : (result | (O_NONBLOCK)));
+#endif
+	DS_EXIT_BLOCKING_PAL_SECTION;
+	return result;
+}
+
+static
+inline
+int
+ipc_poll_fds (
+	ds_ipc_pollfd_t *fds,
+	size_t nfds,
+	uint32_t timeout)
+{
+	int result_poll;
+	DS_ENTER_BLOCKING_PAL_SECTION;
+#ifdef HOST_WIN32
+	result_poll = WSAPoll (fds, (ULONG)nfds, (INT)timeout);
+#else
+	result_poll = poll (fds, nfds, (int)timeout);
```

Comment:
Not just here, but in general around libc calls, should we be checking for EINTR?

Reply:
See comment below.